misk/src/main/kotlin/misk/web/Request.kt

```diff
@@ -1,29 +1,83 @@
 package misk.web
 
 import misk.web.actions.WebSocket
+import misk.web.mediatype.MediaRange
 import okhttp3.Headers
 import okhttp3.HttpUrl
 import okhttp3.MediaType
 import okhttp3.RequestBody
 import okio.BufferedSink
 import okio.BufferedSource
 
-data class Request(
-  val url: HttpUrl,
-  val dispatchMechanism: DispatchMechanism = DispatchMechanism.GET,
-  val headers: Headers = Headers.of(),
-  val body: BufferedSource,
-  val websocket: WebSocket? = null
-) {
-  fun toOkHttp3(): okhttp3.Request {
+/**
+ * A live HTTP call from a client for use by a chain of network interceptors.
+ */
+// TODO(jwilson): rename this class from Request to HttpCall.
+interface Request {
+
+  /** Immutable information about the incoming HTTP request. */
+  val url: HttpUrl
+  val dispatchMechanism: DispatchMechanism
+  val headers: Headers // TODO(jwilson): rename this field to requestHeaders.
+
+  /** The HTTP response under construction. */
+  var statusCode: Int
+  val responseHeaders: Headers
+
+  fun setResponseHeader(name: String, value: String)
+  fun addResponseHeaders(headers: Headers)
+
+  /**
+   * Claim ownership of the request body stream. Returns null if the stream has already been
+   * claimed. Callers should read the HTTP request body or call [withRequestBody] to create a new
+   * chain with a request body that is unclaimed.
+   */
+  fun takeRequestBody(): BufferedSource?
```

Comment:
With feature bindings guaranteeing one and only one handler is taking of the request body, why do we have to provide a nullable type and one claims something already claimed?

Reply:
It’s so we can respond "internal error" if the application code throws, but only if it hasn’t already starting responding